src/data/polynomial/basic.lean

```diff
@@ -35,20 +35,22 @@ variables [semiring R] {p q : polynomial R}
 
 instance : inhabited (polynomial R) := add_monoid_algebra.inhabited _ _
 instance : semiring (polynomial R) := add_monoid_algebra.semiring
-instance : has_scalar R (polynomial R) := add_monoid_algebra.has_scalar
-instance : semimodule R (polynomial R) := add_monoid_algebra.semimodule
+instance {S} [semiring S] [semimodule S R] : semimodule S (polynomial R) :=
+add_monoid_algebra.semimodule
 
 instance [subsingleton R] : unique (polynomial R) := add_monoid_algebra.unique
 
 @[simp] lemma support_zero : (0 : polynomial R).support = ∅ := rfl
 
 /-- `monomial s a` is the monomial `a * X^s` -/
-def monomial (n : ℕ) (a : R) : polynomial R := finsupp.single n a
+def monomial (n : ℕ) : R →ₗ[R] polynomial R := finsupp.lsingle n
```

Comment:
Is there any value in generalizing this to
```suggestion
def monomial {S : Type* := R} [semiring S] [semimodule S R] (n : ℕ) : R →ₗ[S] polynomial R := finsupp.lsingle n
```
I don't know if the opt_param would actually work there.

Reply:
This way we would have to make `S` an explicit parameter. And one can use `linear_map.restrict_scalar` to get this version if needed.